<!-- R Markdown Documentation, DO NOT EDIT THE PLAIN MARKDOWN VERSION OF THIS FILE -->

<!-- Copyright 2014 Google Inc. All rights reserved. -->

<!-- Licensed under the Apache License, Version 2.0 (the "License"); -->
<!-- you may not use this file except in compliance with the License. -->
<!-- You may obtain a copy of the License at -->

<!--     http://www.apache.org/licenses/LICENSE-2.0 -->

<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

Performing QC on gVCF Data using Google Genomics
================================================

The following example makes use of [Illumina Platinum Genomes](http://www.illumina.com/platinumgenomes/).  For more detail about how this data was loaded into the Google Genomics API, please see [Google Genomics Public Data](https://cloud.google.com/genomics/data/platinum-genomes).

Setting Up and Describing the Data
----------------------------------

```{r echo=FALSE, eval=FALSE}
# This codelab assumes that the current working directory is where the Rmd file resides
setwd("/YOUR/PATH/TO/codelabs/R/PlatinumGenomes-QC")
```


```{r comment=NA, message=FALSE, warning=FALSE}
# Setup for BigQuery access
require(bigrquery)
require(xtable)
require(RCurl)
project <- "genomics-public-data"                   # put your projectID here
DisplayAndDispatchQuery <- function(queryUri, replacements=list()) {
  if(grepl("^https.*", queryUri)) {
    querySql <- getURL(queryUri, ssl.verifypeer=FALSE)
  } else {
    querySql <- readChar(queryUri, nchars=1e6)
  }
  for(replacement in names(replacements)) {
    querySql <- gsub(replacement, replacements[[replacement]], querySql, fixed=TRUE)
  }
  cat(querySql)
  query_exec(querySql, project)
}
table_replacement <- list("_THE_TABLE_"="genomics-public-data:platinum_genomes.variants")
```

Let's take a look at a few of the [variants within BRCA1 via BigQuery](https://github.com/googlegenomics/getting-started-bigquery/blob/master/RMarkdown/literate-programming-demo.md#data-visualization)
```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("https://raw.githubusercontent.com/googlegenomics/getting-started-bigquery/master/sql/variant-level-data-for-brca1.sql",
                                  replacements=table_replacement)
```
Number of rows returned by this query: `r nrow(result)`.

Displaying the first few rows of the dataframe of results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```

Notes about this particular dataset:
* It is in gVCF format which adds complexity above and beyond [similar examples for the 1,000 Genomes dataset](https://github.com/googlegenomics/bigquery-examples/blob/master/1000genomes/sql/README.md).
* It is comprised only of SNPs and INDELs (contains no structural variants).
* The values for `alternate_bases` are just comprised of the letters A,C,G,T (e.g., contains no `<DEL>` values).
* It contains some single-allele and 1/2 genotypes.

```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("./sql/characterize-alts.sql",
                                  replacements=table_replacement)
```

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```
We see from the query results that there are no special charaters in alternate_bases and the maximum length is ~50 base pairs.

```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("./sql/genotypes-brca1.sql",
                                  replacements=table_replacement)
```

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```
We see from the query results the variety of genotypes within BRCA1.


Check Singletons
----------------
```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("./sql/private-variants-brca1.sql",
                                  replacements=table_replacement)
```
Number of rows returned by this query: `r nrow(result)`.

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```

Compare to [brca1.singletons](./data/singletons/brca1.singletons) which has 85 some of which are for 0/0 genotypes from reference matching blocks.  This file was created with [vcftools](./data/singletons/brca1.log).

```{r}
require(dplyr)
expectedResult <- read.table("./data/singletons/brca1.singletons", header=TRUE)
# Convert to zero-based coordinates
expectedResult <- mutate(expectedResult, POS = POS - 1)
# Clean colnames to match
colnames(expectedResult) <- gsub('\\.+', '_', colnames(expectedResult))
```

How many singletons do the two results have in common?
```{r}
dim(inner_join(result, expectedResult))
```

Which singletons were only identified by BigQuery?
```{r}
onlyBQ <- anti_join(result, expectedResult)
onlyBQ
```

Which singletons were only identified by vcftools?
```{r}
onlyVcftools <- anti_join(expectedResult, result)
onlyVcftools
```

Retrieving the gVCF data for the singletons identified only by vcftools:
```{r message=FALSE, warning=FALSE, comment=NA}
having <- paste("start = ", onlyVcftools$POS,
                sep="", collapse=" OR ")
result <- DisplayAndDispatchQuery("./sql/examine-data.sql",
                                  replacements=c(table_replacement,
                                                 "_HAVING_"=having))
```

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```

It appears that they correspond either to 
* a reference-matching block, so not actually a singleton and just perhaps violating an assumption in the vcftools code
* or a non-singleon variant, perhaps due to a problem in converting the gVCF data to all-positions VCF via gvcftools?


Check Hardy-Weinberg Equilibrium
-----------------------------------
```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("./sql/hardy-weinberg-brca1.sql",
                                  replacements=table_replacement)
```
Number of rows returned by this query: `r nrow(result)`.

Displaying the first few results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```

Compare to [brca1.hwe](./data/singletons/brca1.hwe).  This file was created with [vcftools](./data/hwe/brca1.log).

```{r}
require(dplyr)
df <- read.table("./data/hwe/brca1.hwe", header=TRUE)
obsSplitCol <- "OBS.HOM1.HET.HOM2."
obsTemp <- read.table(text=as.character(df[, obsSplitCol]), sep = "/")
names(obsTemp) <- c("OBS_HOM1", "OBS_HET", "OBS_HOM2")
eSplitCol <- "E.HOM1.HET.HOM2."
eTemp <- read.table(text=as.character(df[, eSplitCol]), sep = "/")
names(eTemp) <- c("E_HOM1", "E_HET", "E_HOM2")
expectedResult <- cbind(cbind(df[setdiff(names(df), c(obsSplitCol,eSplitCol))], obsTemp), eTemp)
# Convert to zero-based coordinates
expectedResult <- mutate(expectedResult, POS = POS - 1)
```

How many results do the two results have in common?
```{r}
dim(inner_join(result, expectedResult))
```

Which results were only identified by BigQuery?
```{r}
onlyBQ <- anti_join(result, expectedResult, by=c("CHR", "POS"))
onlyBQ
```

Which results were only identified by vcftools?
```{r}
onlyVcftools <- anti_join(expectedResult, result, by=c("CHR", "POS"))
onlyVcftools
```

TODO(deflaux):
* investigate differences in counts
* add Chi-Squared test to query from [this sample](https://github.com/googlegenomics/bigquery-examples/tree/master/1000genomes/data-stories/reproducing-hardy-weinberg-equilibrium)

Note:
(1) we have some single allele genotypes (see query)
(2) vcftools seems to skip those
```
zgrep 41242078 platinum_genomes_brca1_expanded_merged.vcf.gz 
chr17  41242078  .	G	A	143	LowGQX;TruthSensitivityTranche99.90to100.00;LowQD;SiteConflict	BLOCKAVG_min30p3a;MQ=57;MQ0=0;BaseQRankSum=0.781;Dels=0.3;FS=1.561;HRun=11;HaplotypeScore=77.7361;MQRankSum=0.093;QD=2.01;ReadPosRankSum=-2.871;SB=-45.67;VQSLOD=-1.8762;culprit=QD;set=FilteredInAll;DP=425;AF=0.5;AN=25;AC=1	GT:DP:GQX:MQ:AD:GQ:PL:VF	0/0:57:99:59:.:.:.:.	0:27:25:57:26:25.38:.:.	0/0:51:99:57:.:.:.:.	0/1:50:99:59:42,8:99:173,0,1238:0.16	0/0:46:99:59:.:.:.:.	0/0:44:99:60:.:.:.:.	.:46:.:59:40,6:.:.:.	0/0:40:85:59:.:.:.:.	0/0:40:85:59:.:.:.:.	0/0:63:99:58:.:.:.:.	0:42:2:58:37:1.58:.:.	0:33:0:57:29:0.03:.:.	.:44:.:58:31,12:.:.:.	0/0:44:90:58:.:.:.:.	0/0:40:87:58:.:.:.:.	.:44:.:57:39,5:.:.:.	0/0:55:99:59:.:.:.:.
```

Check Individual Heterozygosity
-----------------------------------

More calculations needed here . . . these are just counts.

```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("./sql/homozygous-variants-brca1.sql",
                                  replacements=table_replacement)
```
Number of rows returned by this query: `r nrow(result)`.

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```


